src/forms/form.rs

```diff
@@ -0,0 +1,393 @@
+    pub fn scroll(&mut self, amount: i32) {
+        let mut max_scroll = self.screen.get_height() as i64 - self.get_height() as i64;
+        if self.options.style.border.is_some() {
+            max_scroll += 1;
+        }
+        self.scroll_index =
+            (self.scroll_index as i64 + amount as i64).clamp(0, max_scroll.max(0)) as usize;
+    }
+
+    /// Scroll to a certain position
+    pub fn scroll_to(&mut self, index: usize) {
+        self.scroll(index as i32 - self.scroll_index as i32);
+    }
+
+    /// Adds the provided FormField into the Form
+    ///
+    /// The Field will be resized to match the width of the form  
+    /// You may want to use [build_field](#methods.build_field) instead since it automatically creates the FormField instance.
+    pub fn add_field<T: FormField + 'static>(&mut self, name: &str, mut field: T) {
+        field.resize(self.get_width() - 2, field.get_height());
+        self.fields.push((String::from(name), Box::new(field)));
+    }
+
+    /// Build a field and includes it into the Form
+    ///
+    /// it's an easier approach into building a Form, see the `form-simple` example to compare it with [add_field](#methods.add_field).
+    pub fn build_field<T: FormField + 'static>(&mut self, name: &str, options: FormOptions) {
+        let field = T::make(self.get_width(), options);
+        self.add_field(name, field)
+    }
+
+    /// Get a specific field if it exists within the Form
+    ///
+    /// Note that the field will be removed from the Form
+    pub fn get_field<T>(&mut self, name: &str) -> Option<Box<T>>
+    where
+        T: FormField + 'static,
+    {
+        for (id, (field_name, _)) in self.fields.iter().enumerate() {
+            if name == *field_name {
+                let (_, field) = self.fields.remove(id);
+                return field.to_any().downcast::<T>().ok();
+            }
+        }
+        None
+    }
+
+    /// Get the output of a specific field if it exists and valid within the Form
+    ///
+    /// in case of validation failing, Err value will bear the validation messages directly
+    ///
+    /// See example `form-validation`
+    pub fn get_result(&self, name: &str) -> Result<FormValue, FormError> {
+        for (field_name, field) in self.fields.iter() {
+            if name == *field_name {
+                let mut errors = FormValidationResult::new();
+                field.validate(&mut errors);
+                if errors.is_empty() {
+                    return Ok(field.get_output());
+                } else {
+                    return Err(FormError::ValidationFailed(errors));
+                }
+            }
+        }
+        Err(FormError::FieldNotFound)
+    }
+
+    /// Get the (unvalidated) output of a specific field if it exists within the Form
+    pub fn get_result_unvalidated(&self, name: &str) -> Option<FormValue> {
+        for (field_name, field) in self.fields.iter() {
+            if name == *field_name {
```

Comment:
Since you use this code in a few places, you could abstract it out to a separate function, e.g. `get_field_by_name(&self, name: &str) -> Option<_>`.
Also iterators have `.find` function, so maybe just using that would be enough to cut down the boilerplate code.
But abstracting it out would allow users to access fields.

Now I realize you do have `get_field` function, but it removes the field from the Form. To be fair I don't understand why.
Maybe rename it to `remove_field` and add `get_field` which doesn't remove the field?

Reply:
Actually, I'd like to do this, and that's what `get_field` would do .. I struggled (just a bit... a huge bit in fact 😓) to do it this way and the only solution I was able to make was this removing mess :/

I'll try to see if by any chance this works

Reply:
I guess downcasting is the issue. I don't have much experience with it, so I don't have a solution ready, but I'll try to figure something out.

Reply:
I finally got it to work .. I don't know why it didn't work last time but I'll take it 👍

Reply:
Nice!